tron/utils/state.py

```diff
@@ -1,171 +1,70 @@
-            return False
-
-    def __hash__(self):
-        return hash(self.name)
-
-    def __bool__(self):
-        return bool(self.name)
-
-    def __nonzero__(self):
-        return bool(self.name)
-
-    def __str__(self):
-        return self.name
-
-    def __repr__(self):
-        return "<%r %s>" % (self.__class__.__name__, self.name)
-
-    @property
-    def short_name(self):
-        """If a short_name was given use that, otherwise return the first
-        self._short_letters letters of the name in caps.
-        """
-        if self._short_name:
-            return self._short_name
-        return self.name[:self._short_chars].upper()
-
-
-def traverse(starting_state, match_func):
-    visited = set()
-    state_pairs = [(None, starting_state)]
-
-    def pair_with_name(p):
-        return (p[0], p[1].name)
-
-    while state_pairs:
-        transition_state_pair = state_pairs.pop()
-        _, cur_state = transition_state_pair
-        visited.add(pair_with_name(transition_state_pair))
-
-        if match_func(*transition_state_pair):
-            yield transition_state_pair
-
-        for next_pair in six.iteritems(cur_state):
-            if pair_with_name(next_pair) not in visited:
-                state_pairs.append(next_pair)
-
-
-def named_event_by_name(starting_state, name):
-    name = maybe_decode(name)
-
-    def name_match(t, s):
-        return s.name == name
-
-    try:
-        _, state = next(traverse(starting_state, name_match))
-        return state
-    except StopIteration:
-        raise ValueError("State %s not found." % name)
-
-
-def get_transitions(starting_state):
-    def transition_match(t, s):
-        return bool(t)
-
-    return [trans for trans, _ in traverse(starting_state, transition_match)]
-
-
-class StateMachine(Observable):
-    """StateMachine is a class that can be used for managing state machines.
-
-    A state machine is made up of a State() and a target. The target is the
-    where all the input comes from for making decision about state changes,
-    whatever that may be.
-
-    A State is really just a fancy container for a set of rules for
-    transitioning to other states based on the target.
-    """
-
-    def __init__(self, initial_state, delegate=None, force_state=None):
-        super(StateMachine, self).__init__()
-        self.initial_state = initial_state
-        self.state = force_state or self.initial_state
-        self._state_by_name = None
-        self.delegate = delegate
-        self.transitions = get_transitions(self.initial_state)
+class Machine:
+    @staticmethod
+    def from_machine(machine, initial=None, state=None):
+        if initial is None:
+            initial = machine.initial
+        if state is None:
+            state = initial
+        new_machine = Machine(initial, **machine.transitions)
+        new_machine.state = state
+        assert machine.transitions == new_machine.transitions
+        assert machine.states == new_machine.states
```

Comment:
I don't understand here. Why do we need the assert? What gonna happen if the assert fails?